fw/access_log.c

```diff
@@ -0,0 +1,430 @@
+			chunk++;
+		}
+		break;
+	default:
+		return empty_hdr;
+	}
+	if (chunk == end)
+		return empty_hdr;
+	TFW_STR_INIT(&result);
+	result.len = len;
+	result.chunks = chunk;
+	result.nchunks = end - chunk;
+	return result;
+}
+
+
+void
+do_access_log(TfwHttpResp *resp)
+{
+	do_access_log_req(resp->req, resp->status, resp->content_length);
+}
+
+#define HTTP_METHODS(XX) \
+	XX(COPY)         \
+	XX(DELETE)       \
+	XX(GET)          \
+	XX(HEAD)         \
+	XX(LOCK)         \
+	XX(MKCOL)        \
+	XX(MOVE)         \
+	XX(OPTIONS)      \
+	XX(PATCH)        \
+	XX(POST)         \
+	XX(PROPFIND)     \
+	XX(PROPPATCH)    \
+	XX(PUT)          \
+	XX(TRACE)        \
+	XX(UNLOCK)       \
+	XX(PURGE)
+
+static const struct {
+#	define MAX_HTTP_METHOD_NAME_LEN	10
+	char name[MAX_HTTP_METHOD_NAME_LEN];
+	u8 len;
+} http_methods[] = {
+#define XX(name) [TFW_HTTP_METH_ ## name] = { #name, sizeof(#name) - 1 },
+	HTTP_METHODS(XX)
+#undef XX
+};
+
+static const struct {
+#	define MAX_HTTP_VERSION_LEN 9
+	char name[MAX_HTTP_VERSION_LEN];
+	u8 len;
+} http_versions[] = {
+#define XX(ver, name) [ver] = { name, sizeof(name) - 1 }
+	XX(TFW_HTTP_VER_09, "HTTP/0.9"),
+	XX(TFW_HTTP_VER_10, "HTTP/1.0"),
+	XX(TFW_HTTP_VER_11, "HTTP/1.1"),
+	XX(TFW_HTTP_VER_20, "HTTP/2.0"),
+#undef XX
+};
+
+
+enum {
+#define IGNORE(...)
+#define ENUM(id) idx_ ## id,
+	ACCESS_LOG_LINE(IGNORE, IGNORE, ENUM)
+#undef ENUM
+#undef IGNORE
+	TRUNCATABLE_FIELDS_COUNT
+};
+
+struct str_ref {
+	const char *data;
+	unsigned len;
+};
+
+/** Copies string to buffer if it is not plain. Returns pointer to the
+ * first unused character in the "planar" buffer. */
+static inline char*
+make_plain(char *p, char *end, TfwStr *src, struct str_ref *dst)
+{
+	if (TFW_STR_PLAIN(src)) {
+		dst->data = src->data;
+		dst->len = src->len;
+		return p;
+	} /* protect from BUG_ON in tfw_str_to_cstr */
+	else if (p < end) {
+		dst->len = tfw_str_to_cstr(src, p, end - p);
+		dst->data = p;
+		return p + dst->len;
+	} else {
+		dst->data = "";
+		dst->len = 0;
+		return p;
+	}
+}
+
+/** Truncates truncatable fields if needed. */
+static void
+process_truncated(TfwStr *in, struct str_ref *out, char *p, char *end, 
+		unsigned used_chars)
+{
+	unsigned i;
+	unsigned total_len = 0;
+	unsigned truncated_count = 0;
+	unsigned max_untruncated_len, buf_avail;
+	if (unlikely(used_chars >= ACCESS_LOG_BUF_SIZE))
+		goto no_buffer_space;
+	/* Compute total length of all strings that can be truncated */
+	for (i = 0; i < TRUNCATABLE_FIELDS_COUNT; i++)
+		total_len += in[i].len;
+	/* Check if we're on happy path: all strings fit */
+	if (likely(total_len + used_chars < ACCESS_LOG_BUF_SIZE)) {
+		for (i = 0; i < TRUNCATABLE_FIELDS_COUNT; i++)
+			p = make_plain(p, end, in + i, out + i);
+		return;
+	}
+	/* Unhappy path: evenly distribute available buffer space across all
+	 * strings that do not fit */
+	buf_avail = (ACCESS_LOG_BUF_SIZE - used_chars);
+	max_untruncated_len = buf_avail / TRUNCATABLE_FIELDS_COUNT;
```

Comment:
I still don't like the division, but I do like the flexible truncation much more and I don't see how to avoid the division. It's on the slow path, so I guess we're good with it for now

Reply:
This division could be changed into multiplication by gcc: https://gcc.godbolt.org/z/e1M5ae3hf, but the second one will occur anyway.
But I don't think this would be an issue as we're printing decimal numbers which already involves multiple divisions.

Reply:
This optimization is enabled even on `-Og`. Here's this fragment from `objdump -S` from object file compiled with kernel cflags (which enables -O2):
```
        max_untruncated_len = buf_avail / TRUNCATABLE_FIELDS_COUNT;
 234:   41 b8 ab aa aa aa       mov    $0xaaaaaaab,%r8d
 23a:   89 c7                   mov    %eax,%edi
 23c:   49 0f af f8             imul   %r8,%rdi                    <= multiplication
                if (in[i].len < max_untruncated_len)
 240:   48 c1 ef 21             shr    $0x21,%rdi
 244:   48 39 cf                cmp    %rcx,%rdi
 247:   41 0f 96 c1             setbe  %r9b
                        truncated_count++;
 24b:   48 39 fe                cmp    %rdi,%rsi
 24e:   41 83 d9 ff             sbb    $0xffffffff,%r9d
 252:   48 39 fa                cmp    %rdi,%rdx
 255:   41 83 d9 ff             sbb    $0xffffffff,%r9d
        max_untruncated_len = buf_avail / truncated_count;
 259:   31 d2                   xor    %edx,%edx
 25b:   41 f7 f1                div    %r9d                        <= division
```

Reply:
Ok, cool. Could you please add a comment that sometimes, not always, compiler optimizes devision to multiplication,